snuba/migrations/autogeneration/diff.py

```diff
@@ -0,0 +1,139 @@
+
+    Only supports adding columns, throws error for anything else.
+    """
+    valid, reason = _is_valid_add_column(oldstorage, newstorage)
+    if not valid:
+        raise ValueError(reason)
+
+    oldcol_names = set(
+        col["name"] for col in yaml.safe_load(oldstorage)["schema"]["columns"]
+    )
+    newstorage_dict = yaml.safe_load(newstorage)
+    newcols = newstorage_dict["schema"]["columns"]
+
+    forwardops: list[AddColumn] = []
+    for i, col in enumerate(newcols):
+        if col["name"] not in oldcol_names:
+            column = _schema_column_to_migration_column(parse_columns([col])[0])
+            after = newcols[i - 1]["name"]
+            storage_set = StorageSetKey(newstorage_dict["storage"]["set_key"])
+            forwardops += [
+                AddColumn(
+                    storage_set=storage_set,
+                    table_name=newstorage_dict["schema"]["local_table_name"],
+                    column=column,
+                    after=after,
+                    target=OperationTarget.LOCAL,
+                ),
+                AddColumn(
+                    storage_set=storage_set,
+                    table_name=newstorage_dict["schema"]["dist_table_name"],
+                    column=column,
+                    after=after,
+                    target=OperationTarget.DISTRIBUTED,
+                ),
+            ]
+    return (forwardops, [op.get_reverse() for op in reversed(forwardops)])
+
+
+def _is_valid_add_column(oldstorage: str, newstorage: str) -> tuple[bool, str]:
+    """
+    Input:
+        old_storage, the old storage yaml in str format
+        new_storage, the modified storage yaml in str format
+
+    Returns true if the changes to the storage is valid column addition, false otherwise,
+    along with a reasoning.
+    """
+    oldstorage_dict = yaml.safe_load(oldstorage)
+    newstorage_dict = yaml.safe_load(newstorage)
+    if oldstorage_dict == newstorage_dict:
+        return True, "storages are the same"
+
+    # nothing changed but the columns
+    t1 = oldstorage_dict["schema"].pop("columns")
+    t2 = newstorage_dict["schema"].pop("columns")
+    if not (oldstorage_dict == newstorage_dict):
+        return (
+            False,
+            "Expected the only change to the storage to be the columns, but that is not true",
+        )
+    oldstorage_dict["schema"]["columns"] = t1
+    newstorage_dict["schema"]["columns"] = t2
+
+    # only changes to columns is additions
+    oldstorage_cols = oldstorage_dict["schema"]["columns"]
+    newstorage_cols = newstorage_dict["schema"]["columns"]
+
+    colnames_old = set(e["name"] for e in oldstorage_cols)
+    colnames_new = set(e["name"] for e in newstorage_cols)
+    if not colnames_old.issubset(colnames_new):
+        return (False, "Column removal is not supported")
+
+    pold, pnew = 0, 0
+    while pold < len(oldstorage_cols) and pnew < len(newstorage_cols):
```

Comment:
Kyle, can we just use the set data structure and set operations to determine if new columns are added.

Reply:
Using a set would lose the concept of ordering of columns. Currently the [AddColumn object](https://github.com/getsentry/snuba/blob/271f64e3c026bdb1595fdf2657176174f83ad61a/snuba/migrations/operations.py#L255) has a notion of column order from the "after" parameter. So I used this to make sure the order of the columns matters.